### ISSUE TITLE
Tree Position Resets After Each Build

### ISSUE THREAD

**Issue by JoelProminic** [prominic/Moonshine-IDE#134]
While testing with many projects for #110, I noticed an annoying behavior in the project tree:
1. Open a large project, or multiple projects
2. Expand the project(s) until the scroll bar is visible in the sidebar.  You can open other panels to make this easier
3. Select a file at the bottom of the tree, which is only visible if the sidebar is scrolled to the bottom
4. Project > Build & Run
5. The position of the sidebar resets to the top, so the selected file is no longer visible.

This is a low priority, but it could be annoying in some circumstances.  This does not need to be implemented for the 1.8.0 release.

**Comment by rat-moonshine** [comment]
Tree scroll position now fixed.

**Comment by JoelProminic** [comment]
I checked @rat-moonshine's fix, and the position does not reset to the top for:
* Build
* Build with Apache Ant
* Build & Debug 
* Build as Javascript
* Build & Run as Javascript 

However, it does reset for Clean Project.  We can investigate this after the 1.8.0 release.

**Comment by piotrzarzycki21** [comment]
I believe this is rather bug than enhancement.

**Comment by piotrzarzycki21** [comment]
It also resets after "Delete" file/folder.

**Comment by rat-moonshine** [comment]
It also resets after "Delete" file/folder.

When I tested back, I noticed the selection stays as per the logic developed in case of file/folder deletion, though:
- If the selected file/folder deletes, selection returns to the project root
- If non-selected file/folder deletes, selection stays to the currently selected item

Can you please take a look if these different in your test scenario @piotrzarzycki21 , if yes, please provide test steps to re-produce the problem.

**Comment by rat-moonshine** [comment]
> However, it does reset for Clean Project. We can investigate this after the 1.8.0 release.

In my tests I able to reproduce the above problem in following case:
- When a file/folder is selected inside 'bin' or 'bin-debug' folder and user choose to Clean Project

However, when a file/folder is selected which is not part of the 'bin' or 'bin-debug' folder and user choose to Clean Project, the selection remains to the current file/folder.

**Comment by rat-moonshine** [comment]
I made some changes to the tree selection in case of Clean Project. Please, give a try.

**Comment by piotrzarzycki21** [comment]
Here is what I'm seeing. Maybe that is good behavior, let me know. I'm deleting here folder "obj"

![project_deletion](https://user-images.githubusercontent.com/24554795/35218109-b1debfb8-ff6d-11e7-9f73-8ce287afbca5.gif)

**Comment by rat-moonshine** [comment]
No, this is not right. I confirm the behavior @piotrzarzycki21 . In basic point of view, the 'src' folder suppose to remain open and the file selected inside it (since this isn't the file which has being deleted). I can understand why its happening, I will take a look if this can be handled.

**Comment by rat-moonshine** [comment]
The deletion problem that @piotrzarzycki21 described, that mainly noticed when we add or deletes any file/folder to the project root, and the project root get a refresh call to update the UI. This situation is not when we deletes a file/folder that not exists to the project root.

After some more time spending, this managing became more critical than I initially thought about this. The _TreeView_ tree update codes are very dense and haphazardly distributed through many methods.

So, I went to test in a basic Flex project keeping a Tree component only, and with some basic codes, whether the deletion of item _force_ refresh its parent to update the UI or not. Because if that is so and how Tree component work - then perhaps we would have more tighter option scenarios.

Supplying basic data to the (demo) tree component and deleting the selected node - however, this shown more proper result.

![image](https://user-images.githubusercontent.com/24370870/35271999-c6302eba-0059-11e8-9c39-7a919b0b5656.png)

In above test, deleting "B" node keep the opened leaf items as it is. So, this made clear, something in the TreeView's dense code probably making things something different way. I'm quite tired now at this point searching through that file codes, I shall do that once get some more energy for this.

**Comment by rat-moonshine** [comment]
There were lot of changes undergone to preserve tree states during file/folder addition, deletion and renaming. 

I also observed the project tree resets in case of file/folder creation to the root project, too (the same way which @piotrzarzycki21 reported early). 

Many old methods now removed from using in TreeView now replaced by newer and (probably) the proper codes. Tree state and selection preserving during file/folder deletion/addition/renaming shown more stable results this time during my tests. 

Although, I noticed an old problem when Moonshine starts - selection to the project tree is not displaying effectively all the time. The selection does exists to a particular node but its not showing effectively, a roll-over to the entry reveals the highlight however.

I shall look into that further.
